### ISSUE TITLE
Mac/CoreText: text elements not drawn (or not in the correct colour)

### ISSUE THREAD

**Issue by RJVB** [qtwebkit/qtwebkit#703]
I've run into a site where certain text elements are not being drawn, or are drawn in the wrong (background) colour.

This looked like a CSS parsing issue in OtterBrowser until I discovered that the issue disappears when I select the FreeType font engine.
Edit: it also disappears when I switch OB to the QWE backend.

The screenshots were made with Qt 5.9.3, which did introduce a number of changes to how it renders fonts (but AFAIK more to the Freetype engine than to CoreText).

![screenshot-coretext](https://user-images.githubusercontent.com/1756884/34420364-6c30dcac-ec09-11e7-89e7-037c3b0fc0ce.png)
![screenshot-freetype](https://user-images.githubusercontent.com/1756884/34420370-75704712-ec09-11e7-9a18-426687d621ed.png)

**Comment by annulen** [comment]
Please try reproducing bug in https://github.com/annulen/test_qrawfont - replace font file with disappearing font from the page. Almost certainly this is QtGui bug, not QtWebKit

**Comment by annulen** [comment]
E.g. something like https://github.com/annulen/webkit/issues/452

**Comment by RJVB** [comment]
Hmm, the font in question turns out to be "Open Sans", which I don't have locally. Thanks "WhatFont" GChrome extension!

When I open the font page (https://fonts.google.com/specimen/Open+Sans) I get loads of errors like
```
Failed to compute left/right minimum bearings for "Open Sans"
Failed to compute left/right minimum bearings for "Open Sans Light"
Failed to compute left/right minimum bearings for "Open Sans"
Failed to compute left/right minimum bearings for "Open Sans SemiBold"
Failed to compute left/right minimum bearings for "Open Sans"
Failed to compute left/right minimum bearings for "Open Sans Light"
Failed to compute left/right minimum bearings for "Open Sans"
Failed to compute left/right minimum bearings for "Open Sans SemiBold"
```
on the calling terminal.

Installing the font locally doesn't appear to make any difference to the issue.

Loading it in your test utility does not show any problems: I get a nice "H".

(Could we integrate your utility into my github:RJVB/fontweightissue-qt5.git - e.g. will initialising the QRawFont from a QFont trigger the issue it demonstrates too?)

**Comment by annulen** [comment]
>Hmm, the font in question turns out to be "Open Sans", which I don't have locally. Thanks "WhatFont" GChrome extension!

I just use Inspector to see which font is in the style

>Could we integrate your utility into my github:RJVB/fontweightissue-qt5.git - e.g. will initialising the QRawFont from a QFont trigger the issue it demonstrates too?

Sure! I guess I didn't upload license file there, so terms aren't clear

**Comment by RJVB** [comment]
I cooked up something and pushed it to that repo.
Please have a look if you see anything that you don't agree with (apart from the fact it's very much a first implementation of something I've hardly ever done before).

**Comment by annulen** [comment]
Note that constructing QRawFont from QByteArray and QFont may lead to very different results. QtWebKit uses the former for web fonts and the latter for local.

**Comment by RJVB** [comment]
I guess you'd have to look at the implementation to be certain - I haven't (yet) found a way to determine the font file corresponding to a given QFont.

Either way, I get the same issue in OtterBrowser whether Open Sans is a webfont or a local font.

Also, things get really weird when I open the Inspector and then go into the Styles detail side-bar and start un/re-ticking the various attributes of that .noUI-tooltip element. It's as if the order in which things are drawn changes because varying elements will become invisible or visible again (including the missing text). That only happens with the CoreText fontengine, and I don't see how that can make sense. Unless you're painting over other elements or rather behind them?

**Comment by annulen** [comment]
If you open cppreference.com, can you see missing text like in https://github.com/zealdocs/zeal/issues/776 ?

**Comment by RJVB** [comment]
More exactly, I cannot see the text that's missing  ;) but only in OB; zeal appears to be immune

Interestingly, I had to launch OB with FreeType rendering to be able to get to the actual malloc page. The search hits page below
![screenshot-cppreference](https://user-images.githubusercontent.com/1756884/34436506-b0681d6c-ec96-11e7-9feb-5d02fff9189c.png)

wasn't just missing the text of the links, the links themselves weren't where they should have been either AFAICT. So the text wasn't just invisible as if drawn in the background colour but it's really as if there was something in front of it.

Screenshot with zeal showing normal rendering:
![screenshot-cppreference zeal](https://user-images.githubusercontent.com/1756884/34436531-f2b7e85a-ec96-11e7-90c4-c48e009d4b74.png)

**Comment by RJVB** [comment]
FWIW, I installed "official builds" of both OB and Zeal under Wine on Linux. OB tends to crash on pages known to have rendering problems, but Zeal works just fine. Of course I don't know which rendering engine is used in the Qt builds they include, and I think that ultimately the rendering is handled by the host's FreeType.

**Comment by annulen** [comment]
Are you using HiDPI multiplier?

**Comment by annulen** [comment]
What I actually wanted to know is if bug affects only CoreText or both engines

**Comment by RJVB** [comment]
>Are you using HiDPI multiplier?

No, I only have regular screens, no Hi€€€ ones.

**Comment by annulen** [comment]
That's intertesting, previously it was thought that missing text bug was caused by HiDPI

**Comment by annulen** [comment]
Maybe there is something that works for freetype engine but is broken for all others

**Comment by RJVB** [comment]
Indeed, I'm only seeing it with CoreText, on Mac. The prebuilt Zeal and OB packages don't show it on Win10/64 but you probably checked that yourself already.

I saw on the Zeal issue that the upstream Qt bug is supposedly fixed in Qt 5.9 . I didn't verify, was that a fix that affects the FreeType font engine only, by chance?

**Comment by RJVB** [comment]
OK, could you check my font demonstrator, even try it on MS Windows if you can?

I have now implemented a file picker to create QRawFonts directly from .ttf files, and there's something very strange. Contrary to QRawFonts created from a QFont, those created from file or QByteArray do not advance the cursor when rendering a string with the CoreText font engine. See the screenshots. That's most likely a different issue than missing text, but it shouldn't be happening.

OpenSans-SemiBold.ttf @12pt with CoreText:
![opensanssemibold12pt-coretext](https://user-images.githubusercontent.com/1756884/34446639-0e98558c-ecdd-11e7-98d0-c2c4ce33db74.png)

OpenSans-SemiBold.ttf @12pt with FreeType:
![opensanssemibold12pt-freetype](https://user-images.githubusercontent.com/1756884/34446646-19a8ad5a-ecdd-11e7-9352-08a61a89e00a.png)

**Comment by RJVB** [comment]
Out of curiosity: what does QtWebKit use FontConfig for, and does it do so on all platforms?

**Comment by annulen** [comment]
Fontconfig is used indirectly inside Qt by means of QFontDatabase. There is direct use in testing support code, which is not enabled in production builds. Probably it makes sense only when Qt itself is built with fontconfig, but such check is not preformed

**Comment by annulen** [comment]
>Contrary to QRawFonts created from a QFont, those created from file or QByteArray do not advance the cursor when rendering a string with the CoreText font engine.

I'm not font expert, if you find issues with QRawFont it's better to report them to bugreports.qt.io

**Comment by RJVB** [comment]
>I'm not font expert, if you find issues with QRawFont it's better to report them to bugreports.qt.io

How do you use QRawFonts that you created from a QByteArray? Do you draw individual glyphs or strings?

> Fontconfig is used indirectly inside Qt by means of QFontDatabase.

Which means QtWebKit's reported dependency on it is bogus (esp. on Mac and MS Windows where Qt doesn't use FontConfig). Good to know.

**Comment by annulen** [comment]
>How do you use QRawFonts that you created from a QByteArray? Do you draw individual glyphs or strings?

QPainter::drawGlyphRun is used, like in my test program. For text-stroke QPainterPath is used, but it is not your case probably. See https://code.woboq.org/qt5/qtwebkit/Source/WebCore/platform/graphics/qt/FontCascadeQt.cpp.html#127

**Comment by annulen** [comment]
>Which means QtWebKit's reported dependency on it is bogus

All required dependencies cause cmake error when missing. If you don't enable ENABLE_TEST_SUPPORT, fontconfig should not be even searched for.

**Comment by RJVB** [comment]
>QPainter::drawGlyphRun is used, like in my test program.

I used that code, but instead of filling the QTextLayout with a single QChar I use a QString. I saw there was an override to do that, and assumed that the term "GlyphRun" means the function can handle a run (series) of glyphs. The rest is the same, and the vector with advances I obtain for the text looks correct, but this *is* the first time I'm doing this kind of low-level text rendering in Qt. 
The behaviour inconsistency between the 2 font engines does strike me as a bug but I'd just as well be certain that it's not caused by something that I'm doing wrong.

**Comment by annulen** [comment]
31.12.2017, 00:27, "René Bertin" <notifications@github.com>:
>>QPainter::drawGlyphRun is used, like in my test program.
>
> I used that code, but instead of filling the QTextLayout with a single QChar I use a QString.

Yes, actually QTextLayout is constructed from QString here

https://code.woboq.org/qt5/qtwebkit/Source/WebCore/platform/graphics/qt/FontCascadeQt.cpp.html#179

There is also "simple" code path, it works via https://code.woboq.org/qt5/qtwebkit/Source/WebCore/platform/graphics/qt/FontCascadeQt.cpp.html#292

> I saw there was an override to do that, and assumed that the term "GlyphRun" means the function can handle a run (series) of glyphs.

Yep

> The rest is the same, and the vector with advances I obtain for the text looks correct, but this *is* the first time I'm doing this kind of low-level text rendering in Qt.
> The behaviour inconsistency between the 2 font engines does strike me as a bug but I'd just as well be certain that it's not caused by something that I'm doing wrong.

Probably so

-- 
Regards,
Konstantin

**Comment by annulen** [comment]
Could you try text_layout branch, or applying https://github.com/annulen/webkit/compare/db5bd541403...text_layout ?

**Comment by annulen** [comment]
>Which means QtWebKit's reported dependency on it is bogus (esp. on Mac and MS Windows where Qt doesn't use FontConfig). Good to know.

There is no reported dependency, all dependencies cause cmake fatal error if not found. It is just searched for. I can't reliably detect fontconfig support in all supported Qt versions, and it's technically possible to use it on Mac and Windows, so I look around for it just in case

**Comment by annulen** [comment]
Ah, it must be under ENABLE_TEST_SUPPORT. Sorry

**Comment by annulen** [comment]
I guess this is fixed as a part of #714 (https://bugreports.qt.io/browse/QTBUG-73812)
